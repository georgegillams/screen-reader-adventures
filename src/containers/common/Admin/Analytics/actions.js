import { createRoutine } from 'redux-saga-routines';

import { LOAD } from './constants';

export const load = createRoutine(LOAD);
