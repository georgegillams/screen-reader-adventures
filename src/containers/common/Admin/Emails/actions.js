import { createRoutine } from 'redux-saga-routines';

import { LOAD, RESEND } from './constants';

export const load = createRoutine(LOAD);
export const resend = createRoutine(RESEND);
