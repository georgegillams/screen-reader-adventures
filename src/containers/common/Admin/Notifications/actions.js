import { createRoutine } from 'redux-saga-routines';

import { LOAD, CREATE, REMOVE } from './constants';

export const load = createRoutine(LOAD);
export const create = createRoutine(CREATE);
export const remove = createRoutine(REMOVE);
