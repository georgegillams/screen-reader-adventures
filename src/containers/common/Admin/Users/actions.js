import { createRoutine } from 'redux-saga-routines';

import { LOAD, REMOVE, REQUEST_MAGIC_LINK, CREATE, UPDATE } from './constants';

export const load = createRoutine(LOAD);
export const remove = createRoutine(REMOVE);
export const requestMagicLink = createRoutine(REQUEST_MAGIC_LINK);
export const create = createRoutine(CREATE);
export const update = createRoutine(UPDATE);
