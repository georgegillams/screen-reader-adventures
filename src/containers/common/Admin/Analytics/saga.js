import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load } from './actions';

export function* doLoad() {
  const requestURL = apiStructure.loadAnalyticsSummary.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.analytics));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
}
