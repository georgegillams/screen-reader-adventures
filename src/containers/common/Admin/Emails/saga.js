import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load, resend } from './actions';
import { selectState } from './selectors';

export function* doLoad() {
  const requestURL = apiStructure.loadEmails.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.emails));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export function* doResend() {
  const currentState = yield select(selectState());
  const { emailToResend } = currentState;
  const requestURL = apiStructure.resendEmail.fullPath;

  try {
    yield put(resend.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(emailToResend),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(resend.failure(result));
    } else {
      yield put(resend.success(result.emails));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(resend.failure(err));
  } finally {
    yield put(resend.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
  yield takeLatest(resend.TRIGGER, doResend);
}
