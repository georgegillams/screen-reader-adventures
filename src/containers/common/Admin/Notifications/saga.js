import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { selectState } from './selectors';
import { load, create, remove } from './actions';

export function* doLoad() {
  const requestURL = apiStructure.loadNotifications.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.notifications));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export function* doCreate() {
  const currentState = yield select(selectState());
  const { notificationToCreate } = currentState;
  const requestURL = apiStructure.createNotification.fullPath;

  try {
    yield put(create.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(notificationToCreate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(create.failure(result));
    } else {
      yield put(create.success(result));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(create.failure(err));
  } finally {
    yield put(create.fulfill());
  }
}

export function* doRemove() {
  const currentState = yield select(selectState());
  const { notificationToRemove } = currentState;
  const requestURL = apiStructure.deleteNotification.fullPath;

  try {
    yield put(remove.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(notificationToRemove),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(remove.failure(result));
    } else {
      yield put(remove.success(result));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(remove.failure(err));
  } finally {
    yield put(remove.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
  yield takeLatest(create.TRIGGER, doCreate);
  yield takeLatest(remove.TRIGGER, doRemove);
}
