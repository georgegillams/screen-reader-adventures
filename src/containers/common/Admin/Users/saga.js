import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { selectState } from './selectors';
import { load, remove, requestMagicLink, create, update } from './actions';

export function* doLoad() {
  const requestURL = apiStructure.loadUsers.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.users));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export function* doRemove() {
  const currentState = yield select(selectState());
  const { userToRemove } = currentState;
  const requestURL = apiStructure.deleteUser.fullPath;

  try {
    yield put(remove.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(userToRemove),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(remove.failure(result));
    } else {
      yield put(remove.success(result));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(remove.failure(err));
  } finally {
    yield put(remove.fulfill());
  }
}

export function* doRequestMagicLink() {
  const currentState = yield select(selectState());
  const { userToLogIn } = currentState;
  const requestURL = apiStructure.requestMagicLink.fullPath;

  try {
    yield put(requestMagicLink.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({ email: userToLogIn.email, divertToAdmin: true }),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(requestMagicLink.failure(result));
    } else {
      yield put(requestMagicLink.success(result));
    }
  } catch (err) {
    yield put(requestMagicLink.failure(err));
  } finally {
    yield put(requestMagicLink.fulfill());
  }
}

export function* doCreate() {
  const currentState = yield select(selectState());
  const { userToCreate } = currentState;
  const requestURL = apiStructure.createUser.fullPath;

  try {
    yield put(create.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(userToCreate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(create.failure(result));
    } else {
      yield put(create.success(result));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(create.failure(err));
  } finally {
    yield put(create.fulfill());
  }
}

export function* doUpdate() {
  const currentState = yield select(selectState());
  const { userToUpdate, onUpdateSuccessCb } = currentState;
  const requestURL = apiStructure.updateUser.fullPath;

  try {
    yield put(update.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(userToUpdate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(update.failure(result));
    } else {
      yield put(update.success(result));
      if (onUpdateSuccessCb) {
        onUpdateSuccessCb(result);
      }
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(update.failure(err));
  } finally {
    yield put(update.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
  yield takeLatest(remove.TRIGGER, doRemove);
  yield takeLatest(requestMagicLink.TRIGGER, doRequestMagicLink);
  yield takeLatest(create.TRIGGER, doCreate);
  yield takeLatest(update.TRIGGER, doUpdate);
}
