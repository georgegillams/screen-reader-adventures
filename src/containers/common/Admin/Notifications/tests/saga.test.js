import { put, takeLatest } from 'redux-saga/effects';

import { load, create, remove } from '../actions';

import saga, { doCreate, doLoad, doRemove } from '../saga';

describe('AdminNotifications saga', () => {
  let mainSaga;

  beforeEach(() => {
    mainSaga = saga();
  });

  it('Should load on load TRIGGER', () => {
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(load.TRIGGER, doLoad));
  });

  it('Should create on create TRIGGER', () => {
    mainSaga.next();
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(create.TRIGGER, doCreate));
  });
  it('Should remove on remove TRIGGER', () => {
    mainSaga.next();
    mainSaga.next();
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(remove.TRIGGER, doRemove));
  });

  describe('load actions', () => {
    let loadGenerator;

    const response = {
      notifications: ['notification1', 'notification2'],
      status: 200,
    };

    beforeEach(() => {
      loadGenerator = doLoad();

      const selectDescriptor = loadGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call load.success on successful API call', () => {
      loadGenerator.next();
      const putSuccess = loadGenerator.next(response).value;
      loadGenerator.next();

      expect(putSuccess).toEqual(put(load.success(response.notifications)));
    });

    it('Should call load.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      loadGenerator.next();
      const putFailure = loadGenerator.next(response).value;
      loadGenerator.next();

      expect(putFailure).toEqual(put(load.failure(response)));
    });

    it('Should call load.failure if an exception occurs', () => {
      const response = new Error('Some error');
      const putFailure = loadGenerator.throw(response).value;

      expect(putFailure).toEqual(put(load.failure(response)));
    });
  });

  describe('create actions', () => {
    let createGenerator;

    const response = {
      success: 'Notification created',
      status: 200,
    };

    beforeEach(() => {
      createGenerator = doCreate();

      const selectDescriptor = createGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call create.success on successful API call', () => {
      createGenerator.next({ notificationToCreate: { id: 'u1' } });
      createGenerator.next();
      const putSuccess = createGenerator.next(response).value;
      createGenerator.next();

      expect(putSuccess).toEqual(put(create.success(response)));
    });

    it('Should call load.trigger on successful API call', () => {
      createGenerator.next({ notificationToCreate: { id: 'u1' } });
      createGenerator.next();
      createGenerator.next(response);
      const putSuccess = createGenerator.next().value;

      expect(putSuccess).toEqual(put(load.trigger()));
    });

    it('Should call create.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      createGenerator.next({ notificationToCreate: { id: 'u1' } });
      createGenerator.next();
      const putFailure = createGenerator.next(response).value;
      createGenerator.next(response);

      expect(putFailure).toEqual(put(create.failure(response)));
    });

    it('Should call create.failure if an exception occurs', () => {
      const response = new Error('Some error');
      createGenerator.next({ notificationToCreate: { id: 'u1' } });
      const putFailure = createGenerator.throw(response).value;

      expect(putFailure).toEqual(put(create.failure(response)));
    });
  });

  describe('remove actions', () => {
    let removeGenerator;

    const response = {
      success: 'Notification removed',
      status: 200,
    };

    beforeEach(() => {
      removeGenerator = doRemove();

      const selectDescriptor = removeGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call remove.success on successful API call', () => {
      removeGenerator.next({ notificationToRemove: { id: 'u1' } });
      removeGenerator.next();
      const putSuccess = removeGenerator.next(response).value;
      removeGenerator.next();

      expect(putSuccess).toEqual(put(remove.success(response)));
    });

    it('Should call load.trigger on successful API call', () => {
      removeGenerator.next({ notificationToRemove: { id: 'u1' } });
      removeGenerator.next();
      removeGenerator.next(response);
      const putSuccess = removeGenerator.next().value;

      expect(putSuccess).toEqual(put(load.trigger()));
    });

    it('Should call remove.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      removeGenerator.next({ notificationToRemove: { id: 'u1' } });
      removeGenerator.next();
      const putFailure = removeGenerator.next(response).value;
      removeGenerator.next(response);

      expect(putFailure).toEqual(put(remove.failure(response)));
    });

    it('Should call remove.failure if an exception occurs', () => {
      const response = new Error('Some error');
      removeGenerator.next({ notificationToRemove: { id: 'u1' } });
      const putFailure = removeGenerator.throw(response).value;

      expect(putFailure).toEqual(put(remove.failure(response)));
    });
  });
});
