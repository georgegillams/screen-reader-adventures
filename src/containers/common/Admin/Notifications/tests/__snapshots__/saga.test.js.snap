// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminNotifications saga create actions Should call create.failure if API returns error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`AdminNotifications saga create actions Should call create.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`AdminNotifications saga create actions Should call create.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`AdminNotifications saga create actions Should call load.trigger on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`AdminNotifications saga load actions Should call load.failure if API returns error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/notifications/LOAD/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`AdminNotifications saga load actions Should call load.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/notifications/LOAD/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`AdminNotifications saga load actions Should call load.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/notifications/LOAD/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`AdminNotifications saga remove actions Should call load.trigger on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`AdminNotifications saga remove actions Should call remove.failure if API returns error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`AdminNotifications saga remove actions Should call remove.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`AdminNotifications saga remove actions Should call remove.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
